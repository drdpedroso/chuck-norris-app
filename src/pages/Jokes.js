import React from 'react'
import styled from 'styled-components'
import { getNRandomJokes } from 'services/api/jokesApi'
import { getItemsFromLocalStorage, setItemsInLocalStorage } from 'services/storage/localStorage'
// import { getIntersectionBetweenObjArrays } from 'services/utils/utils'
import JokesList from './jokes/JokesList'
import { mapArrayToHashMap } from '../services/utils/utils'

const JokesWrapper = styled.div`
  display: flex;
`
const Column = styled.div`
  flex-direction: column;
`

class Jokes extends React.Component {
  static getFavoriteJokes() {
    return getItemsFromLocalStorage('jokes') || {}
  }

  static changeFavoriteJokes({favoriteJokes}, joke) {
    const favoriteJokesFormatted = Object.assign(favoriteJokes, {[joke.id]: {...joke, favorite: true}})
    return ({ favoriteJokes: favoriteJokesFormatted })
  }

  static syncJokesBasedOnList(currentState, id, list) {
    debugger
    const selectedJoke = currentState[list][id]
    const newJoke = Object.assign(selectedJoke, {favorite: !selectedJoke.favorite})
    const jokes = Object.assign(currentState[list], {[id]: newJoke})
    return { [list] : jokes }
  }

  constructor() {
    super()
    this.state = {
      jokes: {},
      favoriteJokes: {},
    }

    this.addJokeToFavorites = this.addJokeToFavorites.bind(this)
    this.changeJokeStatus = this.changeJokeStatus.bind(this)
  }

  addJokeToFavorites (joke) {
    this.setState(currentState => Jokes.changeFavoriteJokes(currentState, joke), () =>
      setItemsInLocalStorage('jokes', this.state.favoriteJokes)
    )
  }

  checkForObjectsIntersection(jokes, favoriteJokes) {
    Object.keys(jokes).forEach(joke => {
      if (favoriteJokes[joke]) {
        console.log(joke)
      }
    })
  }

  removeJokeFromFavorites ({id}) {
    this.setState(currentState => {
      const favoriteJokes = Object.assign({}, currentState.favoriteJokes)
      delete favoriteJokes[id]
      return {favoriteJokes}
    }, () =>
      setItemsInLocalStorage('jokes', this.state.favoriteJokes)
    )
  }

  changeJokeStatus (joke, shouldAdd, list) {
    const {id} = joke
    this.setState(currentState => Jokes.syncJokesBasedOnList(currentState, id, list), () => {
      return shouldAdd ? this.addJokeToFavorites(joke) : this.removeJokeFromFavorites(joke)
    })
  }

  async componentDidMount() {
    const { value } = await getNRandomJokes(10)
    const jokes = mapArrayToHashMap(value.map(joke => ({...joke, favorite: false})))
    const favoriteJokes = Jokes.getFavoriteJokes()

    this.setState({ jokes, favoriteJokes }, () => this.checkForObjectsIntersection(this.state.jokes, this.state.favoriteJokes))

  }

  render() {
    const {
      jokes,
      favoriteJokes,
    } = this.state
    return (
      <JokesWrapper>
        <Column>
          <JokesList jokes={jokes} onChange={this.changeJokeStatus} list={'jokes'}/>
        </Column>
        <Column>
          <JokesList jokes={favoriteJokes} onChange={this.changeJokeStatus} list={'favoriteJokes'}/>
        </Column>
      </JokesWrapper>
    )
  }
}

export default Jokes
