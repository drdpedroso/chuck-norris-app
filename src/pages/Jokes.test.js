import React from 'react'
import renderer from 'react-test-renderer'
import { Joke } from 'services/tests/factories'
import Jokes from './Jokes'
import { shallow } from 'enzyme'

describe('Jokes', () => {
  it('should render correctly', () => {
    const component = renderer.create(<Jokes />)
    expect(component).toMatchSnapshot()
  })
  it('should format correctly based on received joke', () => {
    const favoriteJokes ={
      206:{id: 206, joke: "Chuck Norris destroyed the periodic table, because…k Norris only recognizes the element of surprise.", favorite: true},
      223:{id: 223, joke: "Chuck Norris smells what the Rock is cooking... because the Rock is Chuck Norris' personal chef.", favorite: true}
    }
    const res = {
      favoriteJokes: {
        206:{id: 206, joke: "Chuck Norris destroyed the periodic table, because…k Norris only recognizes the element of surprise.", favorite: true},
        223:{id: 223, joke: "Chuck Norris smells what the Rock is cooking... because the Rock is Chuck Norris' personal chef.", favorite: true}
      }
    }
    const joke = {id: 223, joke: "Chuck Norris smells what the Rock is cooking... because the Rock is Chuck Norris' personal chef."}
    const {changeFavoriteJokes} = Jokes
    expect(JSON.stringify(changeFavoriteJokes({favoriteJokes}, joke))).toEqual(JSON.stringify(res))
  })
  it('should sync jokes based on the list that the request was sended', () => {
    const currentState = {
      favoriteJokes: {
        206: {id: 206, joke: "Chuck Norris destroyed the periodic table, because…k Norris only recognizes the element of surprise.", favorite: true}
      },
      jokes: {
        78: {id: 78, joke: "The grass is always greener on the other side, unl…e grass is most likely soaked in blood and tears.", categories: Array(0), favorite: false},
        83: {id: 83, joke: "While urinating, Chuck Norris is easily capable of welding titanium.", categories: Array(0), favorite: false},
        104: {id: 104, joke: "If you Google search &quot;Chuck Norris getting hi…ll generate zero results. It just doesn't happen.", categories: Array(0), favorite: false},
        169: {id: 169, joke: "Chuck Norris won super bowls VII and VIII singleha…tedly retiring to pursue a career in ass-kicking.", categories: Array(0), favorite: false},
        265: {id: 265, joke: "The phrase 'balls to the wall' was originally conc…ing any building smaller than an aircraft hangar.", categories: Array(0), favorite: false},
        329: {id: 329, joke: "There are only two things that can cut diamonds: other diamonds, and Chuck Norris.", categories: Array(0), favorite: false},
        344: {id: 344, joke: "Aliens DO indeed exist. They just know better than to visit a planet that Chuck Norris is on.", categories: Array(0), favorite: false},
        370: {id: 370, joke: "Godzilla is a Japanese rendition of Chuck Norris' first visit to Tokyo.", categories: Array(0), favorite: false},
        402: {id: 402, joke: "There is no such thing as a lesbian, just a woman who has never met Chuck Norris.", categories: Array(0), favorite: false},
        473: {id: 473, joke: "Chuck Norris can overflow your stack just by looking at it.", categories: Array(1), favorite: false},
      }
    }
    const favoriteJokes ={
      206:{id: 206, joke: "Chuck Norris destroyed the periodic table, because…k Norris only recognizes the element of surprise.", favorite: true},
      223:{id: 223, joke: "Chuck Norris smells what the Rock is cooking... because the Rock is Chuck Norris' personal chef.", favorite: true}
    }
    const res = {
      favoriteJokes: {
        206:{id: 206, joke: "Chuck Norris destroyed the periodic table, because…k Norris only recognizes the element of surprise.", favorite: true},
        223:{id: 223, joke: "Chuck Norris smells what the Rock is cooking... because the Rock is Chuck Norris' personal chef.", favorite: true}
      }
    }
    const joke = {id: 223, joke: "Chuck Norris smells what the Rock is cooking... because the Rock is Chuck Norris' personal chef."}
    const {changeFavoriteJokes} = Jokes
    expect(JSON.stringify(changeFavoriteJokes({favoriteJokes}, joke))).toEqual(JSON.stringify(res))
  })
})
