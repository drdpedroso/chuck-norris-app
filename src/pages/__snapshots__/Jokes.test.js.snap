// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jokes should render correctly 1`] = `
<div
  className="sc-bdVaJa kZAgjJ"
>
  <div
    className="sc-bwzfXH dhWCcq"
  />
  <div
    className="sc-bwzfXH dhWCcq"
  />
</div>
`;
